ledger/artifactmanager/artifactmanager_senders.go

```diff
@@ -0,0 +1,156 @@
+/*
+ *    Copyright 2019 Insolar Technologies
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package artifactmanager
+
+import (
+	"context"
+	"sync"
+
+	"github.com/insolar/insolar/core"
+	"github.com/insolar/insolar/core/message"
+	"github.com/insolar/insolar/core/reply"
+	"github.com/insolar/insolar/instrumentation/inslogger"
+	"github.com/insolar/insolar/ledger/storage"
+	"github.com/pkg/errors"
+	"go.opencensus.io/stats"
+)
+
+type ledgerArtifactSenders struct {
+	pulseStorage core.PulseStorage
+	jetStorage   storage.JetStorage
+	defaultBus   core.MessageBus
+
+	codeCacheLock sync.Mutex
+	codeCache     map[core.RecordRef]*cacheEntry
+}
+
+func newLedgerArtifactSenders(
+	pulseStorage core.PulseStorage,
+	jetStorage storage.JetStorage,
+	defaultBus core.MessageBus,
+) *ledgerArtifactSenders {
+	return &ledgerArtifactSenders{
+		pulseStorage: pulseStorage,
+		jetStorage:   jetStorage,
+		defaultBus:   defaultBus,
+		codeCache:    map[core.RecordRef]*cacheEntry{},
+	}
+}
+
+func (m *ledgerArtifactSenders) bus(ctx context.Context) core.MessageBus {
```

Comment:
This should probably be passed as an argument to PreSender constructor. This way, the PreSender won't have to know how to get the bus. We construct PreSenders every time before send, so we can just store MessageBus reference in it and get rid of this method.

Reply:
After conversation there is a decision not to do this action.